File: LeetCode/Java/src/solutions/p0912_SortArray/P0912.java
package solutions.p0912_SortArray;

import java.util.Arrays;

public class P0912 {
    public static void main(String[] args) {

        P0912 s = new P0912();

        int[] nums = {5, 2, 3, 1};
        System.out.println(s.sortArray(nums).toString());

        nums = new int[]{5, 1, 1, 2, 0, 0};
        System.out.println(s.sortArray(nums).toString());

    }

    public int[] sortArray(int[] nums) {
        if (nums.length < 2) {
            return nums;
        }

        int mid = nums.length / 2;

        int[] left = Arrays.copyOfRange(nums, 0, mid);
        int[] right = Arrays.copyOfRange(nums, mid, nums.length);

        return mergeSort(this.sortArray(left), this.sortArray(right));
    }

    public static int[] mergeSort(int[] left, int[] right) {
        int[] result = new int[left.length + right.length];

        int i = 0;
        int l = 0;
        int r = 0;

        while (l < left.length && r < right.length) {
            if (left[l] < right[r]) {
                result[i++] = left[l++];
            } else {
                result[i++] = right[r++];
            }
        }

        if (l < left.length) {
            while (l < left.length) {
                result[i++] = left[l++];
            }
        }

        if (r < right.length) {
            while (r < right.length) {
                result[i++] = right[r++];
            }
        }

        return result;

    }
}

File: LeetCode/Python/solutions/p0912.py
from typing import List


class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:
        #
        # Solution v1: Merge Sort
        #
        # Runtime: 2855 ms @ (beats) 16%
        # Memory Usage: 23.5 MB @ (beats) 23.30%
        #
        if len(nums) < 2:
            return nums

        def merge_sort(left, right):
            result = []
            while left and right:
                if left[0] <= right[0]:
                    result.append(left.pop(0))
                else:
                    result.append(right.pop(0))
            while left:
                result.append(left.pop(0))
            while right:
                result.append(right.pop(0))
            return result

        middle = len(nums) // 2
        left, right = nums[0:middle], nums[middle:]
        return merge_sort(self.sortArray(left), self.sortArray(right))

        #
        # Solution v2: Quick Sort
        #
        # Time Limit Exceeded
        #
        # if len(nums) < 2:
        #     return nums
        #
        # pivot = nums[len(nums) // 2]
        # left = [x for x in nums if x < pivot]
        # mid = [y for y in nums if y == pivot]
        # right = [z for z in nums if z > pivot]
        #
        # return self.sortArray(left) + mid + self.sortArray(right)


if __name__ == "__main__":
    solution = Solution()

    nums = [5, 2, 3, 1]
    print(f"return: {solution.sortArray(nums)}")  # [1,2,3,5]

    nums = [5, 1, 1, 2, 0, 0]
    print(f"return: {solution.sortArray(nums)}")  # [0,0,1,1,2,5]
